pydantic/fields.py

```diff
@@ -563,16 +556,21 @@ def _type_analysis(self) -> None:  # noqa: C901 (ignore complexity)
             return
 
         origin = get_origin(self.type_)
+
+        if origin is Annotated:
+            self.type_ = get_args(self.type_)[0]
+            self._type_analysis()
+            return
+
+        if self.discriminator_key is not None and origin is not Union:
+            raise TypeError('`discriminator` can only be used with `Union` type')
+
```

Comment:
I was using this with a List[Union[]] but obviously I'm running here in this error. Is list union discriminator in scope of this pull-request? The version before 13 april was working as following: 

```
devices: List[
        Union[
            RelaySwitchSchema,
            DBRelaySwitchSchema,
            AM2302Schema,
            BME280Schema,
            SinusSchema,
        ]
    ] = Field(..., discriminator="type")
```

Reply:
I'm so glad when a feature is used beforehand! Allows to have feedback and ensure it works as expected thanks!
Indeed List of Union should be supported but more like `List[MyCustomUnion]` because `discriminator` needs to be added on the `Union` type, not the whole one.
Can you try with
```py
MySchemas = Annotated[
    Union[
        RelaySwitchSchema,
        DBRelaySwitchSchema,
        AM2302Schema,
        BME280Schema,
        SinusSchema,
    ],
    Field(discriminator="type")
]

class MyModel(BaseModel):
    devices: List[MySchemas]
```
?

Reply:
Thanks for your quick reply and the effort you are putting in to this! Your purposed usage of Annotated works like a charm.

Reply:
I'm also in need of exactly this feature, so thanks to the author(s)!

I have a schema setup like the above (list of union), but I have`from_orm` enabled on all models. I receive a traceback when serializing a database model:

```
  File "pydantic/fields.py", line 804, in validate
    v, errors = self._validate_singleton(v, values, loc, cls)
  File "pydantic/fields.py", line 990, in _validate_singleton
    return self._apply_validators(v, values, loc, cls, self.validators)
  File "pydantic/fields.py", line 1034, in _apply_validators
    v = validator(cls, v, values, self, self.model_config)
  File "pydantic/class_validators.py", line 310, in <lambda>
    return lambda cls, v, values, field, config: validator(v)
  File "pydantic/main.py", line 747, in validate
    return cls.from_orm(value)
  File "pydantic/main.py", line 639, in from_orm
    values, fields_set, validation_error = validate_model(cls, obj)
  File "pydantic/main.py", line 1054, in validate_model
    v_, errors_ = field.validate(value, values, loc=field.alias, cls=cls_)
  File "pydantic/fields.py", line 815, in validate
    v, errors = self._validate_sequence_like(v, values, loc, cls)
  File "pydantic/fields.py", line 848, in _validate_sequence_like
    r, ee = self._validate_singleton(v_, values, v_loc, cls)
  File "pydantic/fields.py", line 983, in _validate_singleton
    value, error = field.validate(v, values, loc=loc, cls=cls)
  File "pydantic/fields.py", line 804, in validate
    v, errors = self._validate_singleton(v, values, loc, cls)
  File "pydantic/fields.py", line 990, in _validate_singleton
    return self._apply_validators(v, values, loc, cls, self.validators)
  File "pydantic/fields.py", line 1034, in _apply_validators
    v = validator(cls, v, values, self, self.model_config)
  File "pydantic/class_validators.py", line 310, in <lambda>
    return lambda cls, v, values, field, config: validator(v)
  File "pydantic/main.py", line 747, in validate
    return cls.from_orm(value)
  File "pydantic/main.py", line 639, in from_orm
    values, fields_set, validation_error = validate_model(cls, obj)
  File "pydantic/main.py", line 1054, in validate_model
    v_, errors_ = field.validate(value, values, loc=field.alias, cls=cls_)
  File "pydantic/fields.py", line 815, in validate
    v, errors = self._validate_sequence_like(v, values, loc, cls)
  File "pydantic/fields.py", line 848, in _validate_sequence_like
    r, ee = self._validate_singleton(v_, values, v_loc, cls)
  File "pydantic/fields.py", line 983, in _validate_singleton
    value, error = field.validate(v, values, loc=loc, cls=cls)
  File "pydantic/fields.py", line 804, in validate
    v, errors = self._validate_singleton(v, values, loc, cls)
  File "pydantic/fields.py", line 979, in _validate_singleton
    return self._validate_discriminated_union(v, values, loc, cls)
  File "pydantic/fields.py", line 999, in _validate_discriminated_union
    raise ConfigError(
pydantic.errors.ConfigError: field "_actions" not yet prepared so type is still a ForwardRef, you might need to call EscalationSchema.update_forward_refs().
```

If I actually do the recommendation, I do not receive the traceback and everything is successful.

```python
    @classmethod
    def validate(cls, value):
        cls.update_forward_refs()
        return super().validate(value)
```

Is this expected behavior? I wasn't sure if this was a `from_orm` related bug. I'd be glad to provide more info when I have some time available.

Reply:
I can't reproduce. Could you please provide a full reproduction code or even better a test case to show the error? Thanks

Reply:
Actually, @PrettyWood you have already tested this behavior so it seems as intended behavior to me: https://github.com/samuelcolvin/pydantic/blob/aefcdf1b73a43b43c9b8825aeb3f950ad710300c/tests/test_forward_ref.py#L524